File: android/BiometricPlugin.kt
package dev.wefter.bridge

import android.content.Context
import androidx.biometric.BiometricManager
import androidx.biometric.BiometricManager.Authenticators.BIOMETRIC_STRONG
import androidx.biometric.BiometricManager.Authenticators.BIOMETRIC_WEAK
import androidx.biometric.BiometricManager.Authenticators.DEVICE_CREDENTIAL
import androidx.biometric.BiometricPrompt
import androidx.core.content.ContextCompat
import androidx.fragment.app.FragmentActivity
import org.json.JSONObject

class BiometricPlugin(context: Context, dispatcher: BridgeDispatcher) :
        WefterPlugin(context, dispatcher) {

        private val activity: FragmentActivity
                get() = context as FragmentActivity

        @WefterMethod
        fun isAvailable(payload: JSONObject, callback: (Result<Any>) -> Unit) {
                val allowDeviceCredential = payload.optBoolean("allowDeviceCredential", false)
                val availability =
                        BiometricManager.from(activity)
                                .canAuthenticate(authenticatorsFor(allowDeviceCredential))

                if (availability == BiometricManager.BIOMETRIC_SUCCESS) {
                        resolve(callback, JSONObject().put("available", true))
                        return
                }

                val (code, message) = describeUnavailable(availability)
                resolve(
                        callback,
                        JSONObject()
                                .put("available", false)
                                .put("code", code)
                                .put("message", message)
                )
        }

        @WefterMethod
        fun authenticate(payload: JSONObject, callback: (Result<Any>) -> Unit) {
                val title = payload.optString("title", "Authenticate")
                val subtitle = payload.optString("subtitle", "").takeIf { it.isNotBlank() }
                val cancelText = payload.optString("cancelText", "Cancel")
                val allowDeviceCredential = payload.optBoolean("allowDeviceCredential", false)
                val authenticators = authenticatorsFor(allowDeviceCredential)

                val availability = BiometricManager.from(activity).canAuthenticate(authenticators)
                if (availability != BiometricManager.BIOMETRIC_SUCCESS) {
                        val (code, message) = describeUnavailable(availability)
                        reject(callback, code, message)
                        return
                }

                activity.runOnUiThread {
                        val executor = ContextCompat.getMainExecutor(activity)

                        val promptCallback =
                                object : BiometricPrompt.AuthenticationCallback() {
                                        override fun onAuthenticationSucceeded(
                                                result: BiometricPrompt.AuthenticationResult
                                        ) {
                                                resolve(callback, JSONObject().put("success", true))
                                        }

                                        override fun onAuthenticationError(
                                                errorCode: Int,
                                                errString: CharSequence
                                        ) {
                                                val userInitiated =
                                                        errorCode ==
                                                                BiometricPrompt
                                                                        .ERROR_USER_CANCELED ||
                                                                errorCode ==
                                                                        BiometricPrompt
                                                                                .ERROR_NEGATIVE_BUTTON ||
                                                                errorCode ==
                                                                        BiometricPrompt
                                                                                .ERROR_CANCELED
                                                val lockedOut =
                                                        errorCode ==
                                                                BiometricPrompt.ERROR_LOCKOUT ||
                                                                errorCode ==
                                                                        BiometricPrompt
                                                                                .ERROR_LOCKOUT_PERMANENT

                                                when {
                                                        userInitiated ->
                                                                reject(
                                                                        callback,
                                                                        "USER_CANCELED",
                                                                        "Authentication was canceled"
                                                                )
                                                        lockedOut ->
                                                                reject(
                                                                        callback,
                                                                        "LOCKED_OUT",
                                                                        errString.toString()
                                                                )
                                                        else ->
                                                                reject(
                                                                        callback,
                                                                        "AUTH_ERROR",
                                                                        errString.toString()
                                                                )
                                                }
                                        }

                                        override fun onAuthenticationFailed() {}
                                }

                        val builder =
                                BiometricPrompt.PromptInfo.Builder()
                                        .setTitle(title)
                                        .apply { subtitle?.let { setSubtitle(it) } }
                                        .setAllowedAuthenticators(authenticators)

                        if (!allowDeviceCredential) {
                                builder.setNegativeButtonText(cancelText)
                        }

                        BiometricPrompt(activity, executor, promptCallback)
                                .authenticate(builder.build())
                }
        }

        private fun authenticatorsFor(allowDeviceCredential: Boolean): Int =
                if (allowDeviceCredential) {
                        BIOMETRIC_STRONG or BIOMETRIC_WEAK or DEVICE_CREDENTIAL
                } else {
                        BIOMETRIC_STRONG or BIOMETRIC_WEAK
                }

        private fun describeUnavailable(availability: Int): Pair<String, String> =
                when (availability) {
                        BiometricManager.BIOMETRIC_ERROR_NO_HARDWARE ->
                                "NO_HARDWARE" to "This device has no biometric hardware."
                        BiometricManager.BIOMETRIC_ERROR_HW_UNAVAILABLE ->
                                "HW_UNAVAILABLE" to "Biometric hardware is currently unavailable."
                        BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED ->
                                "NONE_ENROLLED" to
                                        "No fingerprint or face is enrolled on this device."
                        BiometricManager.BIOMETRIC_ERROR_SECURITY_UPDATE_REQUIRED ->
                                "SECURITY_UPDATE_REQUIRED" to
                                        "A security update is required before biometrics can be used."
                        BiometricManager.BIOMETRIC_ERROR_UNSUPPORTED ->
                                "UNSUPPORTED" to
                                        "Biometric authentication is not supported on this device."
                        else -> "UNAVAILABLE" to "Biometric authentication is not available."
                }
}

File: ios/BiometricPlugin.swift
import Foundation
import LocalAuthentication

final class BiometricPlugin: WefterPlugin {

    // @WefterMethod
    func isAvailable(payload: [String: Any], callback: @escaping (Result<Any, Error>) -> Void) throws {
        let allowDeviceCredential = payload["allowDeviceCredential"] as? Bool ?? false
        let context = LAContext()
        let policy: LAPolicy = allowDeviceCredential
            ? .deviceOwnerAuthentication
            : .deviceOwnerAuthenticationWithBiometrics

        var evalError: NSError?
        guard context.canEvaluatePolicy(policy, error: &evalError) else {
            let (code, message) = BiometricPlugin.describeUnavailable(evalError)
            resolve(callback, data: ["available": false, "code": code, "message": message])
            return
        }

        resolve(callback, data: [
            "available": true,
            "biometryType": BiometricPlugin.biometryTypeName(context.biometryType),
        ])
    }

    // @WefterMethod
    func authenticate(payload: [String: Any], callback: @escaping (Result<Any, Error>) -> Void) throws {
        let title = payload["title"] as? String
        let subtitle = payload["subtitle"] as? String
        let reason = subtitle ?? title ?? "Authenticate to continue"
        let allowDeviceCredential = payload["allowDeviceCredential"] as? Bool ?? false

        let context = LAContext()
        if let cancelText = payload["cancelText"] as? String {
            context.localizedCancelTitle = cancelText
        }

        let policy: LAPolicy = allowDeviceCredential
            ? .deviceOwnerAuthentication
            : .deviceOwnerAuthenticationWithBiometrics

        var evalError: NSError?
        guard context.canEvaluatePolicy(policy, error: &evalError) else {
            let (code, message) = BiometricPlugin.describeUnavailable(evalError)
            reject(callback, code: code, message: message)
            return
        }

        context.evaluatePolicy(policy, localizedReason: reason) { [weak self] success, authenticationError in
            guard let self = self else { return }
            if success {
                self.resolve(callback, data: ["success": true])
            } else {
                let (code, message) = BiometricPlugin.describeFailure(authenticationError)
                self.reject(callback, code: code, message: message)
            }
        }
    }

    private static func biometryTypeName(_ type: LABiometryType) -> String {
        switch type {
        case .faceID: return "faceId"
        case .touchID: return "touchId"
        case .opticID: return "opticId"
        default: return "none"
        }
    }

    private static func describeUnavailable(_ error: NSError?) -> (code: String, message: String) {
        guard let laError = error as? LAError else {
            return ("UNAVAILABLE", "Biometric authentication is not available on this device.")
        }

        switch laError.code {
        case .biometryNotAvailable:
            return ("HW_UNAVAILABLE", "Biometric hardware is currently unavailable.")
        case .biometryNotEnrolled:
            return ("NONE_ENROLLED", "No Face ID or Touch ID is enrolled on this device.")
        case .biometryLockout:
            return ("LOCKED_OUT", "Biometric authentication is temporarily locked. Use the device passcode to unlock it.")
        case .passcodeNotSet:
            return ("PASSCODE_NOT_SET", "Set a device passcode to use biometric authentication.")
        default:
            return ("UNAVAILABLE", laError.localizedDescription)
        }
    }

    private static func describeFailure(_ error: Error?) -> (code: String, message: String) {
        guard let laError = error as? LAError else {
            return ("AUTH_ERROR", error?.localizedDescription ?? "Biometric authentication failed.")
        }

        switch laError.code {
        case .userCancel, .systemCancel, .appCancel:
            return ("USER_CANCELED", "Authentication was canceled.")
        case .userFallback:
            return ("USER_CANCELED", "User chose to enter the device passcode instead.")
        case .biometryLockout:
            return ("LOCKED_OUT", "Too many attempts. Biometric authentication is temporarily locked.")
        case .authenticationFailed:
            return ("AUTH_FAILED", "Biometric authentication failed.")
        default:
            return ("AUTH_ERROR", laError.localizedDescription)
        }
    }
}
